checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -363,57 +428,132 @@ private static int getGroupCount(Pattern p) {
     }
 
     /**
-     * Return the non-null groups in the argument.
+     * Returns a list of groups other than 0, that are guaranteed to be non-null given that the
+     * regular expression matches the String. The String argument passed has to be a valid regular
+     * expression.
      *
-     * @param p pattern to be analysed
-     * @return the groups that are non-null in the argument
+     * @param regexp pattern to be analysed
+     * @param n number of capturing groups in the pattern
+     * @return a {@code List} of groups that are guaranteed to match some part of a string that
+     *     matches {@code regexp}
+     * @throws Error if the argument is not a regex
      */
-    private static List<Integer> getNonNullGroups(Pattern p) {
-        String regexp = p.pattern();
-        int n = getGroupCount(p);
+    public static List<Integer> getNonNullGroups(String regexp, int n) {
+        try {
+            Pattern.compile(regexp);
+        } catch (PatternSyntaxException e) {
+            throw new Error(e);
+        }
         List<Integer> nonNullGroups = new ArrayList<>();
-        for (int i = 0; i <= n; i++) {
+        for (int i = 1; i <= n; i++) {
             nonNullGroups.add(i);
         }
+
+        // Holds all indices of opening parentheses that were openings of a capturing group.
         ArrayDeque<Integer> openingIndices = new ArrayDeque<>();
-        int group = 0;
-        boolean squareBracketOpen = false;
+        // Holds whether the last occurrence of a non-literal '(' was a capturing group (true)
+        // or was some other special construct (false). This helps in determining whether a
+        // non-literal
+        // ')' closes a capturing group or some other special construct. If it closes a capturing
+        // group, the top element from openingIndices has to be removed.
+        ArrayDeque<Boolean> openingWasGroup = new ArrayDeque<>();
```

Comment:
Does this Deque have any relationship to the one above, such as that they always have exactly the same length and maybe there is some relationship between elements at corresponding indices?